---
title: "GSE198256_manuscript"
author: "Laura Sudupe. email:laura.medinilla@kaust.eu.sa"
date: "1/31/2024"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
library("NOISeq") #https://rstudio-pubs-static.s3.amazonaws.com/525119_64c1fe6e1a514b89a1ef26d23bf4aae3.html
library("tidyverse")
library("GEOquery")
library("limma")
library("edgeR")
library("clusterProfiler")
library("msigdbr")
library("org.Hs.eg.db")
library("magrittr")
library("EnhancedVolcano")
library("pheatmap")

```

## Introduction

# Overview of the study, dataset origin, and purpose.
The study "Functional reprogramming of monocytes in patients with acute and convalescent severe COVID-19" focuses on understanding the impact of COVID-19 on monocytes, a type of white blood cell crucial for immune response. Using transcriptomic and epigenomic analysis, the study investigates how monocytes are functionally reprogrammed in patients during acute and convalescent phases of severe COVID-19. The dataset GSE198256, derived from this research, offers valuable insights into the changes in monocytes induced by COVID-19.

# Brief explanation of COVID-19 impact on monocytes.
COVID-19, caused by the SARS-CoV-2 virus, has been shown to significantly affect the immune system. Monocytes, as key players in the immune response, undergo functional changes during infection. This study reveals how severe COVID-19 leads to distinctive alterations in cytokine production and gene expression in monocytes. Understanding these changes is critical for comprehending the immune dysregulation observed in severe COVID-19 cases.

# Objectives of the analysis.
The primary objectives of the analysis are:

1. To conduct a comprehensive quality control of the GSE198256 dataset using Noiseq.
2. To perform differential expression analysis, determining how gene expression in monocytes differs between acute COVID-19 patients and healthy individuals.
3. To visualize these differences using volcano plots and heatmaps.
4. To execute Gene Set Enrichment Analysis (GSEA) for a deeper understanding of the biological pathways affected during acute COVID-19 infection.

## Data Acquisition

Description of the GSE198256 dataset.
Steps for downloading and preparing the data for analysis.

```{r data}
# Read data
# load counts table from GEO
urld <- "https://www.ncbi.nlm.nih.gov/geo/download/?format=file&type=rnaseq_counts"
path <- paste(urld, "acc=GSE198256", "file=GSE198256_raw_counts_GRCh38.p13_NCBI.tsv.gz", sep="&");
GSE198256_count <- as.matrix(data.table::fread(path, header=T, colClasses="integer"), rownames=1)

# read metadata
#meta <- read.table("./data/GSE198256/SraRunTable.txt" , sep=",", header = TRUE)
gds <- getGEO("GSE198256")
Meta_GSE198256 <- pData(gds$GSE198256_series_matrix.txt.gz@phenoData)
Meta_GSE198256 <- Meta_GSE198256[,c("title","source_name_ch1","characteristics_ch1",
                                    "characteristics_ch1.1","description","cell type:ch1","disease state:ch1")]

# extract rownames
genes <- rownames(tbl)
#writeLines(genes, "./data/GSE198256/genes_names.txt")

# read biomart export
annotgene <- read.table("./data/GSE198256/mart_export_david.txt",sep = "\t", header = TRUE)
#rownames(annotgene) <- annotgene$Entrezgene # some are repited that could be associated to different isophorms

# How many genes do I get annotated?
sum(rownames(GSE198256_count) %in% annotgene$Entrezgene)

# Filter the information
annotgene <- annotgene[annotgene$Chromosome %in% c(as.character(1:22) ,"X","Y"),]
sum(rownames(GSE198256_count) %in% annotgene$Entrezgene)

## Annotation... solving some issues...
#rownames(annotgene) <- annotgene$Entrezgene
annotgene[annotgene$Entrezgene=="132989",]

annotgene_filt <- annotgene[!duplicated(annotgene$Entrezgene),]
sum(rownames(GSE198256_count) %in% annotgene$Entrezgene)
sum(annotgene_filt$Entrezgene %in% rownames(GSE198256_count))
annotgene_filt[annotgene_filt$Entrezgene=="132989",]

## Overlap between annotation and gnes
rownames(annotgene_filt) <- as.character(annotgene_filt$Entrezgene)
sum(as.character(rownames(annotgene_filt)) %in% rownames(GSE198256_count))

##  Work with the annotated genes!
GSE198256_count_filt <- GSE198256_count[rownames(GSE198256_count) %in% rownames(annotgene_filt),]
GSE198256_count_exc <-GSE198256_count[!(rownames(GSE198256_count) %in% rownames(annotgene_filt)),]
annotgene_ord <- annotgene_filt[rownames(GSE198256_count_filt ),]

sum(rownames(annotgene_ord)==rownames(GSE198256_count_filt))
```


## Quality Control (QC) using Noiseq

# Application of Noiseq to the dataset.

QC in transcriptomic data analysis is a critical step to ensure the reliability and validity of the results. It involves assessing the quality of RNA samples, checking for contamination or degradation, and evaluating the performance of sequencing. In the context of RNA-seq data, Noiseq is a bioinformatics tool used for differential expression analysis in RNA-seq data.


# Presentation and interpretation of QC results.


- Counts per Million (CPM) Distribution Plot: This bar plot showa the distribution of gene expression levels across the samples from the GSE198256 dataset, categorized by different CPM thresholds. A majority of the genes have expression levels above 0 CPM, indicating that they are expressed across the samples. Genes with higher expression levels (CPM > 5 or CPM > 10) are less frequent. This plot provides a snapshot of expression activity within the samples.

- Sequencing Depth vs. Detected Features Plot: This scatter plot shows the relationship between sequencing depth and the number of detected features. As the sequencing depth (million reads) increases, more features are detected, but there's a diminishing return as depth increases. 

- PCA Plot: The PCA plot illustrates the variation among samples by reducing the dimensionality of the dataset to the two principal components (PCs) that explain the most variance. Here, samples are colored based on their condition: Healthy, Covid19: Acute infection, Covid19: Recovery 3Mo, and Covid19: Recovery 6Mo. The plot shows that samples from the 'Covid19: Acute infection' group cluster together, suggesting similar gene expression profiles within this group, while other conditions display more variability.



```{r NOISeq}

Factors_GSE198256 <- data.frame(Meta_GSE198256 [ colnames(GSE198256_count_filt),c("disease state:ch1")])
colnames(Factors_GSE198256)[1]<- "Group"

lengthuse <- abs(annotgene_ord$end-annotgene_ord$start)
names(lengthuse) <- rownames(annotgene_ord)
gc <- annotgene_ord$GC
names(gc) <- rownames(annotgene_ord)
biotype <-annotgene_ord$type
names(biotype) <- rownames(annotgene_ord)

chromosome <- annotgene_ord[,c("Chromosome","start","end")]


data_NOISEQ <- readData(data = GSE198256_count_filt,
                        length=lengthuse,
                        gc=gc,
                        biotype= biotype ,
                        chromosome = annotgene_ord[,c("Chromosome","start","end")],
                        factors = Factors_GSE198256)

myexplodata <- dat(data_NOISEQ, type = "biodetection")
myexplodata <- dat(data_NOISEQ, type = "biodetection")

## explore our data
mysaturation = dat(data_NOISEQ, k = 0, ndepth = 7, type = "saturation")
explo.plot(mysaturation, toplot = 1, samples = 1:8, yleftlim = NULL, yrightlim = NULL)

#sequencing depth
explo.plot(mycountsbio, toplot = 1, samples = NULL, plottype = "barplot")
#pca
myPCA = dat(data_NOISEQ, type = "PCA")
explo.plot(myPCA, factor = "Group")

#for full report run this line
#QCreport(data_NOISEQ, samples = NULL, factor = "Group", norm = FALSE)


## save 
save(data_NOISEQ,GSE198256_count_filt,annotgene_ord,file="./data/GSE198256/GSE198256_step1.Rda")
```





## Differential Expression (DE) Analysis


# Selection and justification of the most appropriate method for this dataset.

For the differential expression (DE) analysis, considering the plots provided for the GSE198256 dataset, the most appropriate method would be limma-voom. This decision is based on the fact that limma-voom is well-suited for datasets where the sequencing depth varies significantly between samples, as it can model the mean-variance relationship of log-counts, effectively handling both lowly and highly expressed genes. Also, this method is robust to the presence of outliers, as showed in the PCA plot, indicating variability among samples.

The focus of our analysis will be on comparing the 'Acute infection' stage with the 'Healthy' control group. This comparison is because the PCA plot indicates a distinct clustering of the acute infection samples, suggesting a strong differential expression signal when compared to the healthy samples. This specific comparison can provide valuable insights into the immune response and molecular changes associated with the acute phase of COVID-19.

```{r Normalize and set design}

# set DGE class
dge <- DGEList(counts=GSE198256_count_filt)

# Make sure on the metadata
rownames(Meta_GSE198256)==colnames(GSE198256_count_filt)
Group[Group=="Covid19: Acute infection"] <- "Covid19AI"
Group[Group=="Covid19: Recovery 3Mo"] <- "Covid193Mo"
Group[Group=="Covid19: Recovery 6Mo"] <- "Covid196Mo"
design <- model.matrix(~ Group )

# Filter
keep <- filterByExpr(dge, design=design)
dge <- dge[keep,,keep.lib.sizes=FALSE]

# Normalization
dge <- calcNormFactors(dge)

```




```{r DE}

## Voom
v <- voom(dge, design, plot=TRUE)
fit_v <- lmFit(v, design)
fit_v <- eBayes(fit_v)
voom_top <-topTable(fit_v, coef=ncol(design))

```

The plot illustrates the mean-variance trend in the dataset. The trend line (in red) shows how the variability of gene expression (y-axis: square root of standard deviation) changes with the average expression level (x-axis: log2(count size + 0.5)). Voom assumes that the variance of log-counts is related to the mean in a specific way: as gene expression increases, the variance tends to decrease. This plot confirms that assumption, showing that highly expressed genes tend to have less variability. The spread of points around the trend line represents individual gene variances and their deviation from the expected trend, important for weighting in subsequent differential expression analysis.


```{r Contrast}

#now I am adding the questions, coeficient are associated now
#to the contrast matrix, no to the design matrix
colnames(design)
contrast.matrix <- makeContrasts(GroupCovid19AI-GroupHealthy, 
                                 GroupCovid196Mo-GroupHealthy,    
                                 levels=design)

fit <- contrasts.fit(fit_v, contrast.matrix)
fit <- eBayes(fit)

topTable(fit) 
topTable(fit,coef=1) #show me the first question results
topTable(fit,coef=2) #compare me all the groups and tell me 
#if there is a difference between all the groups, the F value is
#because of that

# Store all of them
topTable_all <- topTable(fit, number=50)
topTable_q1 <- topTable(fit, coef=1, number=50) # Results for the first contrast
topTable_q2 <- topTable(fit, coef=2, number=50) # Results for the second contrast

```


## Visualization: Volcano Plot and Heatmap

Creating and interpreting a volcano plot for DE genes.
Generating a heatmap for acute infection and healthy comparisons.

```{r annotation}
## annotate the genes
gene_symbols <- mapIds(org.Hs.eg.db,
                       keys = rownames(topTable_q1), # your gene IDs
                       column = "SYMBOL",
                       keytype = "ENTREZID",
                       multiVals = "first")

rownames(topTable_q1) <- gene_symbols[rownames(topTable_q1)]

```



```{r volcano and heatmap}

## volcano
EnhancedVolcano(topTable_q1,
                lab = rownames(topTable_q1),
                x = 'logFC', # This is the log2 fold change column in your topTable
                y = 'P.Value', # This is the p-value column in your topTable
                xlim = c(-10, 10),
                title = "Volcano plot: Acute vs Healthy")


## heatmap
# Extract the group information from the 'characteristics_ch1.1' column
Meta_GSE198256$Group <- gsub("disease state: ", "", Meta_GSE198256$characteristics_ch1.1)

# Create the sample_info dataframe with the relevant group information
sample_info <- data.frame(Group = Meta_GSE198256$Group, row.names = colnames(dge$counts))

# Ensure that annotation_col has the same row names as the count data for correct matching
annotation_col <- as.data.frame(sample_info$Group, row.names = rownames(sample_info))

# Select the top 20 genes based on the absolute log-fold change from the topTable results
topGenes <- rownames(topTable_q1[order(abs(topTable_q1$logFC), decreasing = TRUE), ][1:30, ])

# Extract the normalized logCPM counts for the top genes
normalized_logCPM <- cpm(dge, log=TRUE, prior.count=5)
topGenes_data <- normalized_logCPM[topGenes,]

# Generate the heatmap using pheatmap
pheatmap(topGenes_data,
         annotation_col = annotation_col,
         show_rownames = TRUE,
         show_colnames = TRUE,
         fontsize_row = 6,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         main = "Top 30 genes: Acute vs Healthy")


```

In the volcano plot, each point represents a gene. The x-axis shows the log2 fold change, indicating the magnitude and direction of expression changes, while the y-axis shows the negative log10 of the p-value, indicating the statistical significance. Genes that appear toward the top of the plot are more statistically significant, while those further to the left or right have greater fold changes. The red points are likely set to highlight genes that meet certain thresholds for fold change and significance, suggesting they are of particular interest.

The heatmap, shows  the expression patterns of the top 30 differentially expressed genes across different samples grouped by their condition (Acute infection, Recovery 3Mo, Recovery 6Mo, Healthy). The color scale indicates the expression level, with red showing higher expression and blue indicating lower expression.



## Gene Set Enrichment Analysis (GSEA)

Rationale for choosing GSEA over ORA (Over-Representation Analysis).
Execution of GSEA for acute infection samples.
Analysis and interpretation of pathway enrichment results.

```{r GSEA}

# Step 1: determine genes of interest.
diff_table_all <- topTable(fit2,coef=1,p.value=1,number=nrow(logCPM)) 

# Step 2: determine background.

# Step 3: Determine gene sets.

msigdbr_species()
hs_msigdb_df <- msigdbr(species = "Homo sapiens")
head(hs_msigdb_df)

hs_kegg_df <- hs_msigdb_df %>%
  dplyr::filter(
    gs_cat == "C2", # This is to filter only to the C2 curated gene sets
    gs_subcat == "CP:KEGG" # This is because we only want KEGG pathways
  )

# Step 4: conduct GSEA

list_ordered <- diff_table_all[,"B"]
names(list_ordered) <- rownames(diff_table_all)
  
  
gsea_results <- GSEA(
  geneList = list_ordered, # Ordered ranked gene list
  minGSSize = 25, # Minimum gene set size
  maxGSSize = 500, # Maximum gene set set
  pvalueCutoff = 0.05, # p-value cutoff
  eps = 0, # Boundary for calculating the p value
  seed = 123, # Set seed to make results reproducible
  pAdjustMethod = "BH", # Benjamini-Hochberg correction
  TERM2GENE = dplyr::select(
    hs_kegg_df,
    gs_name,
    human_entrez_gene
  )
)



# Step 5: Visualize / explore


head(gsea_results@result)

gsea_result_df <- data.frame(gsea_results@result)
gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(NES, n = 3)

most_positive_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "KEGG_JAK_STAT_SIGNALING_PATHWAY",
  title = "KEGG_JAK_STAT_SIGNALING_PATHWAY",
  color.line = "#0d76ff"
)
most_positive_nes_plot

gsea_result_df %>%
  # Return the 3 rows with the smallest (most negative) NES values
  dplyr::slice_min(NES, n = 3)

most_negative_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "KEGG_SPLICEOSOME",
  title = "KEGG_SPLICEOSOME",
  color.line = "#0d76ff"
)
most_negative_nes_plot


```


## Conclusion

Summary of key findings and their implications.
Reflection on the methodology and potential future research directions.
